components/dialog/dialog-mixin.js

```diff
@@ -231,12 +231,32 @@ export const DialogMixin = superclass => class extends RtlMixin(superclass) {
 		if (this._opener && this._opener.focus) {
 			// wait for inactive focus trap
 			requestAnimationFrame(() => {
-				tryApplyFocus(this._opener);
+				let focusable = this._opener;
 				this._opener = null;
+				if (!isFocusable(focusable)) {
+					focusable = findComposedAncestor(focusable, (node) => (isFocusable(node) || getFirstFocusableDescendant(node) !== null));
+					if (focusable === null) return;
+					if (!isFocusable(focusable)) {
+						focusable = getFirstFocusableDescendant(focusable);
+					}
+				}
+				if (isFocusable(focusable)) focusable.focus();
 			});
 		}
 	}
 
+	_forceDialogFocusVisible(elem) {
+		if (!isFocusable(elem, false, false)) {
+			elem = getFirstFocusableDescendant(elem);
+		}
+		if (!elem) return;
+		if (elem.tagName === 'BUTTON') {
```

Comment:
Theoretically this could end up putting the CSS class on random buttons that have no idea about it, but I think that's OK? I originally had it looking for `<d2l-button>`, but sometimes this got called with its inner `<button>` so it was too late to check for it.